dask/array/linalg.py

```diff
@@ -48,14 +48,32 @@ def tsqr(data, name=None, compute_svd=False, _max_vchunk_size=None):
     Singular Value Decomposition.  It requires that the input array have a
     single column of blocks, each of which fit in memory.
 
-    If blocks are of size ``(n, k)`` then this algorithm has memory use that
-    scales as ``n**2 * k * nthreads``.
-
     Parameters
     ----------
     data: Array
     compute_svd: bool
         Whether to compute the SVD rather than the QR decomposition
+    use_recursion: bool
+        Whether to use recursion
+
+    Notes
+    ----------
+    If blocks are of size ``(n, k)`` then this algorithm has memory use that
+    scales as ``n**2 * k * nthreads``.
+
+    The implementation here is the recursive variant due to the need for a
+    "single core" QR decomposition. Given m blocks, the size of the required
+    "single core" QR decomposition will have to work with a ``(m * k, k)``
+    matrix, and ``m * k`` may be substantially larger than ``n``.
+
+    As such, if ``n / k >= gamma``, recursion will be applied as necessary
+    (unless prevented with the ``use_recursion`` flag) to ensure that single
+    core computations do not work on blocks larger than ``(n, k)``.
+
+    Where blocks are irregular, the above logic is applied with the "height" of
+    the "tallest" block used in place of ``n``.
```

Comment:
Is gamma defined somewhere?

This block starts off talking about how "the implementation here is the recursive variant due ..." which, while is something we care about as developers, may not be as important to users as what to expect.  Instead, I recommend starting off with a statement like the following:

> For an ``(n, m)`` array cut into ``nb`` number of ``(k, m)`` blocks, this algorithm uses XXX memory. 

This answers what I expect users will care about most upfront.  We might follow it up with details about the recursion if they're important for users or with notes about rechunking.

Reply:
Damn. I was in a hurry. It was 2. I'm setting it back to 2.

Reply:
Speaking of rechunking, if you're up for it it might be interesting to add a `rechunk=True` keyword argument to this algorithm and others that performs automatic rechunking for the user.  This would require some thought about what rechunking would be optimal.

This would be future work though, it'd be good to get this in.  See also https://github.com/dask/dask/issues/3587

Reply:
Rechunking would be interesting. My current views are that for the most part the user should try to manage that... I like the `rechunk='10MB'` type of option.

Reply:
But yes. "Future work."

Reply:
Regardless it would be useful to get your thoughts on how a user should best rechunk their data when using this algorithm (or sfqr).  

Say we're trying to optimize for low memory use first, and short compute times second, if I have an N by M array and prefer chunk sizes that are less than C bytes, how should my array be chunked if I want to use this algorithm?

Reply:
As tall as possible. That reduces the number of nodes in the tree. So auto-chunking is not a problem here. Just specify a size and auto-chunking works intuitively.